### DISCUSSION TITLE
FXGL 17 Roadmap

### DISCUSSION THREAD

**Discussion by AlmasB** [AlmasB/FXGL#1089] [General]
Hi all,

This is the FXGL 17 discussion space. It would be good to agree a few key goals for the FXGL 17 release (in 2022), which will change its baseline Java and target to 17, as well as use JavaFX 17.

There are still various open issues that could be made into goals, but given ever-rising interest in Raspberry Pi support, I think this should be a priority. If agreed, some technical help may be needed from @FDelporte @DieterHolz @johanvos 

As always, suggestions and ideas are very welcome.

Also CC to community developers involved with FXGL projects:

@adambocco 
@carldea
@ImperaEtConquer
@chengenzhao
@dykstrom

**Comment by FDelporte** [comment]
Would it be possible to add an "embedded" label to the related open issues?

FPS improvements/bottle necks is definitely something to investigate for FXGL on Raspberry Pi. 
@DieterHolz I don't know if this could be the subject of a student project at FHNW?

**Reply by DieterHolz** [reply]
We have "small" student projects and the bachelor thesis.

Both are done by teams of two students.

In the small projects every student should work approximately 180 hours, in the bachelor thesis 360 hours.

I would assume that FPS improvements are not enough work for 2x180 hours. 

We can put it in a larger project that covers "everything you need to use FXGL on Pi" for example with a new default game-menu, substitution of all default mouse-events (you don't have a mouse on Pi when using an arcade console), etc.

**Reply by AlmasB** [reply]
@FDelporte added a new "embedded" label.
@DieterHolz happy to talk potential projects. 

I'm happy to try and fix the FPS issue, but I currently don't have a clue why it occurs. The quickest way to diagnose this is probably a chat over Teams or Zoom, so we can debug live. @FDelporte and I should be able to do this. What do you think, Frank?

**Reply by FDelporte** [reply]
Thanks for the label! I'm taking holiday the next three weeks, so trying to stay away from computer ;-) Happy to do a call afterwards to do some initial research on Raspberry Pi performance!

**Comment by chengenzhao** [comment]
In terms of embedded system and linux 
Steam Deck seems like going to release soon
so is it possible to export FXGL game to Steam Deck which uses Steam OS, a Linux distro
and theoretically Java, JavaFX and FXGL could be very easy to export to Steam Deck 
just needs some work on Steam SDK
and I think this could be a very interesting subject for student works.
e.g. Steam SDK for FXGL

**Reply by FDelporte** [reply]
SteamOS on Raspberry Pi doesn't seem to be well supported yet? Only found this "recent" message (4/2020): https://steamcommunity.com/groups/steamuniverse/discussions/0/2259061617890416251/

**Reply by chengenzhao** [reply]
Hmm.... seems like Raspberry Pi doesn't support yet
But if we could have Steam SDK for FXGL
then other hardwares could also support FXGL games

**Comment by Higgins328** [comment]
It would be nice to have the opportunity to embed FXGL apps into an existing Java FX application
I can see that there was an issue rised for this in 2019. Not that it an issue as I see it, but it sure would have been useful

https://github.com/AlmasB/FXGL/issues/611

**Reply by FDelporte** [reply]
Is possible! 😃
https://twitter.com/AlmasBaim/status/1417446341702193152?s=19

**Comment by chengenzhao** [comment]
Just read the tweet https://twitter.com/AlmasBaim/status/1423310378415673350
since 3D is not experimental any more
it would be nice to have a way combining different 2D and 3D sub scenes in the same scene e.g. game scene
JavaFX provides a way of mixing 2D and 3D contents ([JavaFX subscene class](https://docs.oracle.com/javase/8/javafx/api/javafx/scene/SubScene.html))
but FXGL hasn't support it yet
so maybe this feature is a nice to have in 17?

more discussion: https://github.com/AlmasB/FXGL/discussions/1057

**Reply by chengenzhao** [reply]
I just watch the video about HD-2D
https://www.youtube.com/watch?v=jCZzoFJfMpI
If the 2D and 3D could mix together 
JavaFX and FXGL maybe useful for this scenario
